bindings/profilers/wall.cc

```diff
@@ -970,29 +1033,127 @@ v8::CpuProfiler* WallProfiler::CreateV8CpuProfiler() {
   return cpuProfiler_;
 }
 
-v8::Local<v8::Value> WallProfiler::GetContext(Isolate* isolate) {
-  auto context = *curContext_.load(std::memory_order_relaxed);
-  if (!context) return v8::Undefined(isolate);
-  return context->Get(isolate);
+Local<Value> WallProfiler::GetContext(Isolate* isolate) {
+  auto context = GetContextPtr(isolate);
+  if (context) {
+    return context->Get(isolate);
+  }
+  return Undefined(isolate);
 }
 
 void WallProfiler::SetContext(Isolate* isolate, Local<Value> value) {
-  // Need to be careful here, because we might be interrupted by a
-  // signal handler that will make use of curContext_.
-  // Update of shared_ptr is not atomic, so instead we use a pointer
-  // (curContext_) that points on two shared_ptr (context1_ and context2_),
-  // update the shared_ptr that is not currently in use and then atomically
-  // update curContext_.
-  auto newCurContext = curContext_.load(std::memory_order_relaxed) == &context1_
-                           ? &context2_
-                           : &context1_;
-  if (!value->IsNullOrUndefined()) {
-    *newCurContext = std::make_shared<Global<Value>>(isolate, value);
+#if NODE_MAJOR_VERSION >= 23
+  if (!useCPED_) {
+    curContext_.Set(isolate, value);
+    return;
+  }
+
+  // Clean up dead context pointers
+  for (auto it = deadContextPtrs_.begin(); it != deadContextPtrs_.end();) {
+    auto ptr = *it;
+    liveContextPtrs_.erase(ptr);
+    delete ptr;
+    ++it;
+  }
+  deadContextPtrs_.clear();
+
+  auto cped = isolate->GetContinuationPreservedEmbedderData();
+  // No Node AsyncContextFrame in this continuation yet
+  if (!cped->IsObject()) return;
+
+  auto v8Ctx = isolate->GetCurrentContext();
+  // This should always be called from a V8 context, but check just in case.
+  if (v8Ctx.IsEmpty()) return;
+
+  auto cpedObj = cped.As<Object>();
+  auto localSymbol = cpedSymbol_.Get(isolate);
+  auto maybeProfData = cpedObj->Get(v8Ctx, localSymbol);
+  if (maybeProfData.IsEmpty()) return;
+
+  PersistentContextPtr* contextPtr = nullptr;
+  auto profData = maybeProfData.ToLocalChecked();
+  if (profData->IsUndefined()) {
+    contextPtr = new PersistentContextPtr(&deadContextPtrs_);
```

Comment:
could this be too many new operations ?

Reply:
There will be at most as many as there are `AsyncContextFrame` (ACF) objects in Node (practically, there'll be exactly as many.) With GC callbacks we'll keep them deleted in sync with how ACFs are being GCd in V8 heap. It's theoretically possible that V8 could still create new ACF instances on its heap (as it allocates bigger chunks of memory up front) but we can't allocate a C++ object on the heap. I would like to think it is very unlikely. We might want to add comments here in the code though to explain this.